### DISCUSSION TITLE
Really long production build time of typescript app on Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by mimoid-prog** [vercel/next.js#17709] [Help] [answered]
# Bug report

Next.js apps with typescript have really long production build time or even the build never ends if the app is bigger. 
It takes 5 minutes to build my simple typescript app with one page and my other much bigger app can't even finish the build process. (It also takes 100% of server CPU usage)
The most time it takes on "Creating an optimized production build" step.
The problem appears only on the **Ubuntu** server on Digital Ocean. The same apps built on my **Windows** system do the build really quickly like in 10-60sec depending on how big is the app.

## To Reproduce

Try to clone my repo, pretty simple typescript app with custom server and material ui:
https://github.com/Ptak-Warsaw-Expo/test-expoplus 

Then try to build the app on Ubuntu 20.04 on Digital Ocean.

## Expected behavior

Much faster build time. The same app without typescript (or on Windows) builds x10 times faster.

## System information

- OS: Ubuntu 20.04
- VPS: Digital Ocean droplet
- Version of Next.js: 9.5.3
- Version of Node.js: 14.10.1

**Comment by mattcarlotta** [comment, marked as answer]
This has nothing to do with Next but a limitation of using a cheap VPS. In short, the cheaper VPS, the less CPU power and less RAM will be allocated to it. For example, their $5 VPS is assigned 1 CPU core and 1GB of RAM and their $10 VPS is 1 CPU core and 2GB of RAM. Unfortunately, neither option is even (or barely) enough to install dependencies, let alone build an application. 

The recommended solution would be build locally and transfer the build via rsync or FTP. You can do this manually or you can set up a CI (using GitHub Actions) to do it [automatically](https://github.com/mattcarlotta/SJSITApp-SSR/blob/master/.github/workflows/deployment.yml).

**Reply by mimoid-prog** [reply]
You were right. I upgraded to 15$ VPS and everything started to work. And thanks for the advice with local builds

**Reply by aryomuzakki** [reply]
> You were right. I upgraded to 15$ VPS and everything started to work. And thanks for the advice with local builds

 @mimoid-prog 

I'm sorry, could you please describe more about your 15$ VPS ?
How many CPU core and RAM was needed for your build ?

**Reply by mattcarlotta** [reply]
@aryomuzakki Go to [Digital Ocean's pricing page](https://www.digitalocean.com/pricing) and select the **Droplets (virtual machines)** offerings and select **Regular with SSD,** which should give you 2vCPUs/2GB RAM. This may not be enough if your application is larger/has a lot of pages; if so, I'd recommend the $20  2vCPUs/4GB RAM droplet -- or find yourself a hosting service that offers similar specs for a VPS.
![Screenshot from 2022-03-27 09-30-17](https://user-images.githubusercontent.com/22607722/160291464-ddea3ff1-60c3-40fd-b578-4100efa42ef4.png)

Ideally, to save money, you can build locally or build using Github Actions, and then send the build over to your droplet -- you just need to make sure both systems are running the same OS and are utilizing the same Node version.

**Reply by aryomuzakki** [reply]
@mattcarlotta sorry I haven't replied yet, I wasn't aware because I never use their services. Thank you very much for your explanation and advice.

**Reply by bozzhik** [reply]
Actually, this is not always due to a bad VPS, you just need to go through the whole project again. In my case, tailwind scanned all the files, including `/node_modules` and the build took several hours and did not complete successfully, now it is 136 sec

https://github.com/vercel/next.js/discussions/60147#discussioncomment-7998187

![image](https://github.com/user-attachments/assets/054e1d97-cf00-4e5c-ae6a-1074f19aea9c)

Deployments: 

![image](https://github.com/user-attachments/assets/2309d859-33f4-4a88-a50d-65c6bddfc2b3)
